enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,339 @@
+---
+title: AWS EFS CSI driver operator via OLM
+authors:
+  - "@jsafrane"
+reviewers:
+  - 2uasimojo # from OCP Dedicated team (i.e. "the customer")
+  - bertinatto # from OCP storage
+approvers:
+  - bbennett # As pillar lead?
+
+creation-date: 2021-03-09
+last-updated: 2021-06-24
+status: implementable
+see-also:
+  - "/enhancements/storage/csi-driver-install.md"
+replaces:
+superseded-by:
+---
+
+# AWS EFS CSI driver operator via OLM
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Operational readiness criteria is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+AWS EFS CSI driver (+ corresponding operator) is an optional OCP component. It should be installed through OLM, when
+users opts-in.
+
+This document describes existing (unsupported) solution and how to turn it into a supported one, both from code (and
+build and shipment), and from user perspective.
+
+## Motivation
+
+Some of our customers want to use Amazon Elastic FileSystem in their clusters. At the same time, not all customers want
+to use it, so it should not be installed by default as our AWS EBS CSI driver.
+
+Despite written by Red Hat employees, the existing CSI driver operator
+[aws-efs-operator](https://github.com/openshift/aws-efs-operator) uses upstream CSI driver, upstream CSI sidecars and
+[the operator itself](https://quay.io/repository/app-sre/aws-efs-operator?tab=tags) is not released through
+registry.redhat.com and therefore cannot be supported by us.
+
+### Goals
+
+* Allow users to use AWS EFS volumes in a supported way.
+* Allow existing users of community-supported aws-efs-operator to update to Red Hat supported solution.
+* Keep existing community community-supported aws-efs-operator features (namely, `SharedVolume` in
+  `aws-efs.managed.openshift.io` API group). See below.
```

Comment:
IMO this should be a non-goal. Problems with SharedVolume in the existing operator:
- It is insecure (TL;DR: anyone with permissions to create SharedVolume can attach to any access point)
- Customers don't like the automatic creation of PVCs, and particularly the strict naming convention.
- No support for dynamic provisioning.

In order to keep SharedVolume, you would need to change it so much that it might as well be a different CR, or none at all.

Another soft concern is customers not understanding whether their SharedVolume is being managed by the old thing (insecure, unsupported) or the new thing (secure, supported). If they think it's the latter but it's really the former, they could end up pretty upset.

Reply:
Thanks for bringing this up, I am +1 for dropping this requirement. Besides making things easier for us, the current `SharedVolume` feels a bit "out of place" compared to other solutions.

Reply:
Removed SharedVolume as a separate commit.

Reply:
Still seeing this bullet in the latest commit.

Reply:
I think it's correct now. We won't remove the community-supported aws-efs-operator from Operator Hub. Users may choose the old one or not to upgrade to the new one, if they depend on `SharedVolume`.

Reply:
Okay, the wording was confusing to me, made it still sound like the *new* solution was going to support `SharedVolume`. It's made clear further down, so no big deal.